README.md
# API Development and Documentation Final Project

## Trivia App

Udacity is invested in creating bonding experiences for its employees and students. A bunch of team members got the idea to hold trivia on a regular basis and created a webpage to manage the trivia app and play the game, but their API experience is limited and still needs to be built out.

That's where you come in! Help them finish the trivia app so they can start holding trivia and seeing who's the most knowledgeable of the bunch. The application must:

1. Display questions - both all questions and by category. Questions should show the question, category and difficulty rating by default and can show/hide the answer.
2. Delete questions.
3. Add questions and require that they include question and answer text.
4. Search for questions based on a text query string.
5. Play the quiz game, randomizing either all questions or within a specific category.

Completing this trivia app will give you the ability to structure plan, implement, and test an API - skills essential for enabling your future applications to communicate with others.

## Starting and Submitting the Project

[Fork](https://help.github.com/en/articles/fork-a-repo) the project repository and [clone](https://help.github.com/en/articles/cloning-a-repository) your forked repository to your machine. Work on the project locally and make sure to push all your changes to the remote repository before submitting the link to your repository in the Classroom.

## About the Stack

We started the full stack application for you. It is designed with some key functional areas:

### Backend

The [backend](./backend/README.md) directory contains a partially completed Flask and SQLAlchemy server. You will work primarily in `__init__.py` to define your endpoints and can reference models.py for DB and SQLAlchemy setup. These are the files you'd want to edit in the backend:

1. `backend/flaskr/__init__.py`
2. `backend/test_flaskr.py`

> View the [Backend README](./backend/README.md) for more details.

### Frontend

The [frontend](./frontend/README.md) directory contains a complete React frontend to consume the data from the Flask server. If you have prior experience building a frontend application, you should feel free to edit the endpoints as you see fit for the backend you design. If you do not have prior experience building a frontend application, you should read through the frontend code before starting and make notes regarding:

1. What are the end points and HTTP methods the frontend is expecting to consume?
2. How are the requests from the frontend formatted? Are they expecting certain parameters or payloads?

Pay special attention to what data the frontend is expecting from each API response to help guide how you format your API. These are the files you'd want to edit in the frontend:

1. `frontend/src/components/QuestionView.js`
2. `frontend/src/components/FormView.js`
3. `frontend/src/components/QuizView.js`

By making notes ahead of time, you will practice the core skill of being able to read and understand code and will have a simple plan to follow to build out the endpoints of your backend API.

> View the [Frontend README](./frontend/README.md) for more details.


backend/CHANGELOG.md

# Changelog
Created this file.

## INITIAL SETUP

### Install Dependencies
DONE beforehand. Note: Not using Virtual Environment.

### Set up Database
DID:
```bash
    sudo service postgresql start
    sudo -u postgres -i
        createdb trivia
    sudo -u postgres psql
        \c trivia
        \i trivia.psql
```
DID:
```
    updated `database_path` in models.py
    updated ../.gitignore to exclude `backend/models.py` (for security reasons)
        + cd ..
        + git update-index --skip-worktree backend/models.py
```

### Run the Server
DID:
```bash
    export FLASK_APP=flaskr
    export FLASK_ENV=development
    flask run --reload
```

### Tried Testing
DID:
```bash
    python3 test_flaskr.py
```

## FIRST APP ROUTE WITH TESTING

### Implemented app route = GET /categories
DID updates to `__init__.py`:
```
    CORS(app)
    @app.after_request
    @app.route("/categories")
```

### Set up Test Database
DID:
```bash
    sudo -u postgres -i
        createdb trivia_test
    sudo -u postgres psql
        \c trivia_test
        \i trivia.psql
```

### Implemented test for GET /categories
DID updates to `test_flaskr.py`:
```
    updated `database_path`
    test_retrieve_categories
```
DID:
```bash
    python3 -W ignore::DeprecationWarning test_flaskr.py
```
RESULT:
```
.
----------------------------------------------------------------------
Ran 1 test in 0.037s

OK
```

### Updated bakend/README.md
DID:
    Documenting your Endpoints
    Testing (note about deprecation warnings)

## SECOND APP ROUTE

### Implemented app route = GET /questions?page=#
DID updates to `__init__.py`:
```
    app.config['JSON_SORT_KEYS'] = False # we decide the order
    def paginate_questions(request, questions)
    @app.route("/questions")
```
DID updates to `test_flaskr.py`:
```
    test_retrieve_questions
```

## REMAINING APP ROUTES

### Testing
Created `re_insert_questions.sql` for re-establishing deleted questions as part of testing!

backend/test_flaskr.py
import os
import unittest
import json
from flask_sqlalchemy import SQLAlchemy

from flaskr import create_app
from models import setup_db, Question, Category


class TriviaTestCase(unittest.TestCase):
    """This class represents the trivia test case"""

    def setUp(self):
        """Define test variables and initialize app."""
        self.app = create_app()
        self.client = self.app.test_client
        self.database_name = "trivia_test"
        self.database_path = "postgresql://{}:{}@{}/{}".format('postgres','<password>','localhost:5432', self.database_name)
        setup_db(self.app, self.database_path)

        # binds the app to the current context
        with self.app.app_context():
            self.db = SQLAlchemy()
            self.db.init_app(self.app)
            # create all tables
            self.db.create_all()
    
    def tearDown(self):
        """Executed after each test"""
        pass

    def test_retrieve_categories(self):
        """Test GET categories"""
        res = self.client().get("/categories")
        data = json.loads(res.data)
        self.assertEqual(res.status_code, 200)
        self.assertTrue(data["categories"])
        self.assertGreaterEqual(len(data["categories"]), 2) # expect more than one category

    def test_retrieve_questions(self):
        """Test GET questions, with optional page parameter"""
        # No page parameter
        res = self.client().get("/questions")
        data = json.loads(res.data)
        self.assertEqual(res.status_code, 200)
        self.assertTrue(data["questions"])
        self.assertLessEqual(len(data["questions"]), 10) # max 10 questions per page
        self.assertGreaterEqual(len(data["questions"]), 1) # at least one question (otherwise it would be an error)
        self.assertTrue(data["totalQuestions"])
        self.assertTrue(data["categories"])
        self.assertGreaterEqual(len(data["categories"]), 2)
        self.assertTrue(data["currentCategory"])
        # Page 1 - should be same as above (defaulting to page=1)
        res = self.client().get("/questions?page=1")
        data1 = json.loads(res.data)
        self.assertEqual(res.status_code, 200)
        self.assertEqual(data, data1) # expect same content (for page 1) as without specifying page parameter
        # Page 2
        res = self.client().get("/questions?page=2")
        data2 = json.loads(res.data)
        self.assertEqual(res.status_code, 200)
        self.assertNotEqual(data2, data1) # expect different content (for page 2) from page 1
        self.assertTrue(data2["questions"])
        self.assertLessEqual(len(data2["questions"]), 10) # max 10 questions per page
        self.assertGreaterEqual(len(data2["questions"]), 1) # at least one question (otherwise it would be an error)
        self.assertTrue(data2["totalQuestions"])
        self.assertTrue(data2["categories"])
        self.assertGreaterEqual(len(data2["categories"]), 2)
        self.assertTrue(data2["currentCategory"])
        # Page 999 (non-existent)
        res = self.client().get("/questions?page=999")
        self.assertEqual(res.status_code, 404) # expecting not to find any questions on this far out page

    def test_retrieve_category_questions(self):
        """Test GET category questions"""
        # Questions for Category 1
        res = self.client().get("/categories/1/questions")
        data = json.loads(res.data)
        self.assertEqual(res.status_code, 200)
        self.assertTrue(data["questions"])
        self.assertTrue(data["totalQuestions"])
        self.assertTrue(data["currentCategory"])
        # Category 999 (non-existent)
        res = self.client().get("/categories/999/questions")
        self.assertEqual(res.status_code, 404)

    def test_delete_question(self):
        """Test DELETE question"""
        # Delete an exising question
        res = self.client().delete("/questions/5")
        self.assertEqual(res.status_code, 200)
        # Deleting the same question again - should result in error (showing that the previous attempt succeeded)
        res = self.client().delete("/questions/5")
        self.assertEqual(res.status_code, 422)

    def test_create_question(self):
        """Test create question"""
        # Create a new question
        new_question = {"question": "Is this the new answer string?", "answer": "Yes", "difficulty": 1, "category": 3}
        res = self.client().post("/questions", json=new_question)
        self.assertEqual(res.status_code, 200)
        # @TODO: Add more asserts, etc.

    def test_search_questions(self):
        """Test searching questions"""
        # Search for some questions - knowing the result will be non-empty
        res = self.client().post("/questions", json={"searchTerm": "title"})
        self.assertEqual(res.status_code, 200)
        # @TODO: Add more asserts, etc.
        # Search for some questions - knowing the result will be empty
        res = self.client().post("/questions", json={"searchTerm": "non_existing_term"})
        self.assertEqual(res.status_code, 200)
        # @TODO: Add more asserts, etc.
        # Invalid request body
        res = self.client().post("/questions", json={"searchForThis": "title"})
        self.assertEqual(res.status_code, 422)

    def test_retrieve_next_question(self):
        """Test retrieving next question (in quizz)"""
        # Ask for next question in Entertainment category
        next_question = { 'previus_questions': [2, 4], 'quiz_category': 'Entertainment' }
        res = self.client().post("/quizzes", json=next_question)
        self.assertEqual(res.status_code, 200)
        # @TODO: Add more asserts, etc.

# Make the tests conveniently executable
if __name__ == "__main__":
    unittest.main()


frontend/src/components/Question.js
import React, { Component } from 'react';
import '../stylesheets/Question.css';

class Question extends Component {
  constructor() {
    super();
    this.state = {
      visibleAnswer: false,
    };
  }

  flipVisibility() {
    this.setState({ visibleAnswer: !this.state.visibleAnswer });
  }

  render() {
    const { question, answer, category, difficulty } = this.props;
    return (
      <div className='Question-holder'>
        <div className='Question'>{question}</div>
        <div className='Question-status'>
          <img
            className='category'
            alt={`${category.toLowerCase()}`}
            src={`${category.toLowerCase()}.svg`}
          />
          <div className='difficulty'>Difficulty: {difficulty}</div>
          <img
            src='delete.png'
            alt='delete'
            className='delete clickable'
            onClick={() => this.props.questionAction('DELETE')}
          />
        </div>
        <div
          className='show-answer button clickable'
          onClick={() => this.flipVisibility()}
        >
          {this.state.visibleAnswer ? 'Hide' : 'Show'} Answer
        </div>
        <div className='answer-holder'>
          <span
            style={{
              visibility: this.state.visibleAnswer ? 'visible' : 'hidden',
            }}
          >
            Answer: {answer}
          </span>
        </div>
      </div>
    );
  }
}

export default Question;


frontend/src/components/QuestionView.js
import React, { Component } from 'react';
import '../stylesheets/App.css';
import Question from './Question';
import Search from './Search';
import $ from 'jquery';

class QuestionView extends Component {
  constructor() {
    super();
    this.state = {
      questions: [],
      page: 1,
      totalQuestions: 0,
      categories: {},
      currentCategory: null,
    };
  }

  componentDidMount() {
    this.getQuestions();
  }

  getQuestions = () => {
    $.ajax({
      url: `/questions?page=${this.state.page}`, //TODO: update request URL
      type: 'GET',
      success: (result) => {
        this.setState({
          questions: result.questions,
          totalQuestions: result.totalQuestions,
          categories: result.categories,
          currentCategory: result.currentCategory,
        });
        return;
      },
      error: (error) => {
        alert('Unable to load questions. Please try your request again');
        return;
      },
    });
  };

  selectPage(num) {
    this.setState({ page: num }, () => this.getQuestions());
  }

  createPagination() {
    let pageNumbers = [];
    let maxPage = Math.ceil(this.state.totalQuestions / 10);
    for (let i = 1; i <= maxPage; i++) {
      pageNumbers.push(
        <span
          key={i}
          className={`page-num clickable ${i === this.state.page ? 'active' : ''}`}
          onClick={() => {
            this.selectPage(i);
          }}
        >
          {i}
        </span>
      );
    }
    return pageNumbers;
  }

  getByCategory = (id) => {
    $.ajax({
      url: `/categories/${id}/questions`, //TODO: update request URL
      type: 'GET',
      success: (result) => {
        this.setState({
          questions: result.questions,
          totalQuestions: result.total_questions,
          currentCategory: result.current_category,
        });
        return;
      },
      error: (error) => {
        alert('Unable to load questions. Please try your request again');
        return;
      },
    });
  };

  submitSearch = (searchTerm) => {
    $.ajax({
      url: `/questions`, //TODO: update request URL
      type: 'POST',
      dataType: 'json',
      contentType: 'application/json',
      data: JSON.stringify({ searchTerm: searchTerm }),
      xhrFields: {
        withCredentials: true,
      },
      crossDomain: true,
      success: (result) => {
        this.setState({
          questions: result.questions,
          totalQuestions: result.total_questions,
          currentCategory: result.current_category,
        });
        return;
      },
      error: (error) => {
        alert('Unable to load questions. Please try your request again');
        return;
      },
    });
  };

  questionAction = (id) => (action) => {
    if (action === 'DELETE') {
      if (window.confirm('are you sure you want to delete the question?')) {
        $.ajax({
          url: `/questions/${id}`, //TODO: update request URL
          type: 'DELETE',
          success: (result) => {
            this.getQuestions();
          },
          error: (error) => {
            alert('Unable to load questions. Please try your request again');
            return;
          },
        });
      }
    }
  };

  render() {
    return (
      <div className='question-view'>
        <div className='categories-list'>
          <h2
            onClick={() => {
              this.getQuestions();
            }}
          >
            Categories
          </h2>
          <ul>
            {Object.keys(this.state.categories).map((id) => (
              <li
                key={id}
                onClick={() => {
                  this.getByCategory(id);
                }}
                className='clickable'
              >
                {this.state.categories[id]}
                <img
                  className='category'
                  alt={`${this.state.categories[id].toLowerCase()}`}
                  src={`${this.state.categories[id].toLowerCase()}.svg`}
                />
              </li>
            ))}
          </ul>
          <Search submitSearch={this.submitSearch} />
        </div>
        <div className='questions-list'>
          <h2>Questions</h2>
          {this.state.questions.map((q, ind) => (
            <Question
              key={q.id}
              question={q.question}
              answer={q.answer}
              category={this.state.categories[q.category]}
              difficulty={q.difficulty}
              questionAction={this.questionAction(q.id)}
            />
          ))}
          <div className='pagination-menu'>{this.createPagination()}</div>
        </div>
      </div>
    );
  }
}

export default QuestionView;


backend/flaskr/__init__.py
import os
from flask import Flask, request, abort, jsonify
from flask_sqlalchemy import SQLAlchemy
from flask_cors import CORS
import random

from models import setup_db, Question, Category

QUESTIONS_PER_PAGE = 10

def paginate_questions(request, questions):
    page = request.args.get("page", 1, type=int)
    start_page_question = (page - 1) * QUESTIONS_PER_PAGE
    end_page_question = start_page_question + QUESTIONS_PER_PAGE

    all_questions = [question.format() for question in questions]
    page_questions = all_questions[start_page_question:end_page_question]

    return page_questions


def create_app(test_config=None):
    # create and configure the app
    app = Flask(__name__)
    app.config['JSON_SORT_KEYS'] = False # we decide the order
    setup_db(app)
    CORS(app)

    @app.after_request
    def after_request(response):
        response.headers.add(
            "Access-Control-Allow-Headers", "Content-Type,Authorization,true"
        )
        response.headers.add(
            "Access-Control-Allow-Methods", "GET,POST,DELETE" # consider including PUT and OPTIONS
        )
        return response

    @app.route("/categories")
    def retrieve_categories():
        """
        Endpoint to handle GET requests for all available categories.
        
        GET '/categories'
        {
          'categories': { 
            '1' : "Science",
            '2' : "Art",
            '3' : "Geography",
            '4' : "History",
            '5' : "Entertainment",
            '6' : "Sports"
          }
        }
        """
        categories = Category.query.order_by(Category.id).all()
        data={}
        for category in categories:
            data[category.id]=category.type
        if len(data) == 0:
            abort(404)
        else:
            return jsonify(
                {
                    "categories": data,
                }
            )

    @app.route("/questions")
    def retrieve_questions():
        """
        Endpoint to handle GET requests for questions, including pagination (every 10 questions).
        This endpoint returns a list of questions, number of total questions, current category, categories.
        
        GET '/questions?page=${integer}'
        {
          'questions': [
            {
              'id': 1,
              'question': 'This is a question',
              'answer': 'This is an answer',
              'difficulty': 5,
              'category': 2
            },
            ...
          ],
          'totalQuestions': 100,
          'categories': { 
            '1' : "Science",
            ...
          },
          'currentCategory': 'History'
        }
        """
        questions = Question.query.order_by(Question.category).order_by(Question.id).all()
        page_questions = paginate_questions(request, questions)
        if len(page_questions) == 0:
            abort(404)
        else:
            categories = Category.query.order_by(Category.id).all()
            category_data={}
            for category in categories:
                category_data[category.id]=category.type

            return jsonify(
                {
                    "questions": page_questions,
                    "totalQuestions": len(Question.query.all()),
                    "categories": category_data,
                    "currentCategory": category_data[int(page_questions[-1:][0]['category'])]
                }
            )

    @app.route('/questions/<int:question_id>', methods=["DELETE"])
    def delete_question(question_id):
        """
        Endpoint to DELETE question using a question ID.
        Deletes a specified question using the id of the question
        Request Arguments: id - integer
        Returns: Does not need to return anything besides the appropriate HTTP status code.
        This removal will persist in the database and when you refresh the page.

        DELETE '/questions/${id}'
        """
        try:
            question=Question.query.filter(Question.id==question_id).one_or_none()
            if question is None:
                abort(404)
            else:
                question.delete()
                return (jsonify({"success": True}), 200)
        except:
            abort(422)

    def create_question(question, answer, difficulty, category):
        """
        NB: Called from create_question_or_search_questions "shared" app route /questions - see below

        Endpoint to POST a new question, which will require the question and answer text, category, and difficulty score.
        Sends a post request in order to add a new question
        Returns: Does not return any new data

        POST '/questions'
        request body:
        {
            'question':  'Heres a new question string',
            'answer':  'Heres a new answer string',
            'difficulty': 1,
            'category': 3,
        }
        """
        try:
            question=Question(question, answer, category, difficulty)
            question.insert()
            return (jsonify({"success": True}), 200)
        except:
            abort(422)

    def search_questions(search_term, category):
        """
        NB: Called from create_question_or_search_questions "shared" app route /questions - see below

        Endpoint to get questions based on a search term via a POST.
        It returns any questions for whom the search term is a substring of the question.
        Sends a post request in order to search for a specific question by search term
        Returns: any array of questions, a number of totalQuestions that met the search term and the current category string
        
        POST '/questions'
        request body:
        {
            'searchTerm': 'this is the term the user is looking for'
        }
        reply:
        {
            'questions': [
                {
                    'id': 1,
                    'question': 'This is a question',
                    'answer': 'This is an answer',
                    'difficulty': 5,
                    'category': 5
                },
                ...
            ],
            'totalQuestions': 1,
            'currentCategory': 'Entertainment'
        }
        """
        questions=Question.query.order_by(Question.category).order_by(Question.id).filter(Question.question.ilike('%{}%'.format(search_term))).all()
        page_questions=paginate_questions(request, questions)
        total_questions=len(questions)
        current_category=category
        return jsonify(
            {
                'questions': page_questions,
                'totalQuestions': total_questions,
                'currentCategory': current_category
            }
        )
    
    @app.route('/questions', methods=["POST"])
    def create_question_or_search_questions():
        """
        Shared endpoint (app route) for create_question or search_questions - see above
        """
        body = request.get_json()
        search_term = body.get("searchTerm", None)
        if search_term is None:
            # Create question
            new_question = body.get("question", None)
            new_answer = body.get("answer", None)
            new_difficulty = body.get("difficulty", None)
            new_category = body.get("category", None)
            if new_question is None or new_answer is None or new_difficulty is None or new_category is None:
                abort(422)
            else:
                return create_question(new_question, new_answer, new_difficulty, new_category)
        else:
            # Search questions
            category = body.get("category", None)
            return search_questions(search_term, category)

    @app.route('/categories/<category_id>/questions')
    def retrieve_category_questions(category_id):
        """
        Endpoint to get questions based on category.
        Fetches questions for a cateogry specified by id request argument
        Request Arguments: id - integer
        Returns: An object with questions for the specified category, total questions, and current category string

        GET '/categories/${id}/questions'
        {
          'questions': [
            {
              'id': 1,
              'question': 'This is a question',
              'answer': 'This is an answer',
              'difficulty': 5,
              'category': 4
            },
            ...
          ],
          'totalQuestions': 100,
          'currentCategory': 'History'
        }
        """
        category=Category.query.filter(Category.id==int(category_id)).one_or_none()
        if category is None:
            abort(404)
        else:
            questions=Question.query.filter(Question.category==category_id).order_by(Question.id).all()
            all_questions = [question.format() for question in questions]
            return jsonify(
                {
                    "questions": all_questions,
                    "totalQuestions": len(Question.query.all()),
                    "currentCategory": category.type
                }
            )

    @app.route('/quizzes', methods=["POST"])
    def retrieve_next_question():
        """
        Endpoint to get questions to play the quiz.
        This endpoint takes category and previous question parameters and return a random questions within the given category, if provided, and that is not one of the previous questions.
        Sends a post request in order to get the next question
        Returns: a single new question object

        POST '/quizzes'
        request body:
        {
            'previous_questions': [1, 4, 20, 15],
            'quiz_category': {
                'type': category_name, 
                'id': category_id
            }
        }
        reply:
        {
            'question': {
                'id': 1,
                'question': 'This is a question',
                'answer': 'This is an answer',
                'difficulty': 5,
                'category': 4
            }
        }
        """
        body = request.get_json()
        previous_questions = body.get("previous_questions", None)
        quiz_category = body.get("quiz_category", None)
        if quiz_category is None:
            abort(422)
        else:
            try:
                category_id = quiz_category['id']
                if category_id == 0: # ALL
                    question = Question.query.filter(Question.id.not_in(previous_questions)).first()
                else:
                    question = Question.query.filter(Question.category==str(category_id)).filter(Question.id.not_in(previous_questions)).first()
                return jsonify(
                    {
                        'question': question.format()
                    }
                )
            except:
                abort(422)

    """
    Error handlers for all expected errors - including 404 and 422.
    """
    @app.errorhandler(404)
    def not_found(error):
        return (
            jsonify({"success": False, "error": 404, "message": "Resource not found"}),
            404
        )

    @app.errorhandler(422)
    def unprocessable(error):
        return (
            jsonify({"success": False, "error": 422, "message": "Unable to process request"}),
            422
        )

    @app.errorhandler(400)
    def bad_request(error):
        return (
            jsonify({"success": False, "error": 400, "message": "Bad request"}), 
            400
        )

    return app
